src/tokenizer.rs

```diff
@@ -396,14 +398,59 @@ fn parse_string_literal<'a>(current: &'a str, current_position: &SourcePosition)
     Ok((advance_result, TokenKind::StringLiteral(literal)))
 }
 
-fn parse_long_string_literal<'a>(current: &'a str, current_position: &SourcePosition) -> Result<(AdvanceResult<'a>, TokenKind<'a>), AdvanceError> {
+fn parse_multi_line_string_literal<'a>(current: &'a str, current_position: &SourcePosition) -> Result<(AdvanceResult<'a>, TokenKind<'a>), AdvanceError> {
+    match parse_multi_line_thing(current, &current_position) {
+        Err(e) => Err(e),
+        Ok((advance_result, raw_content, depth)) =>
+            Ok((advance_result, TokenKind::StringLiteral(
+                StringLiteral::LongForm { raw_content, depth },
+            )))
+    }
+}
+
+fn parse_multi_line_thing<'a>(current: &'a str, current_position: &SourcePosition) -> Result<(AdvanceResult<'a>, Cow<'a, str>, u32), AdvanceError> {
+    let captures = match PATTERN_MULTI_LINE_STRING_START.captures(current) {
+        Some(c) => c,
+        None => return Err(AdvanceError::NoMatch),
+    };
+
+    let start_match = captures.get(0).unwrap();
+
+    let rest = &current[start_match.end()..];
+    let depth = captures.get(1).unwrap().as_str().len() as u32;
+
+    let reg_str = format!(r"]={{{}}}]", depth); // Expands to something like ]={5}]
+    let end_reg = Regex::new(reg_str.as_str()).unwrap();
+
+    let end_captures = match end_reg.captures(rest) {
+        Some(em) => em,
+        None => return Err(AdvanceError::Error(TokenizeError::UnclosedString {
+            position: *current_position,
+        })),
+    };
```

Comment:
We could probably transform this to `.ok_or_else(|| ...)?` if you think it would be cleaner.

I'm slooowly learning how to apply Rust's `Option<->Result` operations and the question mark operator for great good.

Reply:
`.ok_or()?` works great here too.